modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,304 @@
+        self.__has_ran_once = False
+    
+    def clear_all_data(self):
+        """
+        Deletes all current points stored and emptys the current bucket
+        """
+        self.__all_points = []
+        self.__current_bucket = []
+        self.__has_ran_once = False
+
+    def reset_model(self):
+        """
+        Resets model back to original initialization, before any data has been fitted
+        """
+        self.__vgmm_model = sklearn.base.clone(self.__vgmm_model)
+
+    def run(self, detections: "list[DetectionInWorld]", run_override: bool) -> "tuple[bool, list[ObjectInWorld | None]]":
+        """
+        Take in list of landing pad detections and return list of estimated landing pad locations
+        if number of detections is sufficient, or if manually forced to run.
+
+        PARAMETERS
+        ----------
+        detections: list[DetectionInWorld]
+            list containing DetectionInWorld objects containing real-world positioning data to run clustering on
+        
+        run_override: bool, default = False 
+            forces ClusterEstimation worker to run and output predictions regardless of other conditions
+
+        RETURNS
+        -------
+        model_ran: bool
+            True if ClusterEstimation worker ran its estimation model
+
+        objects_in_world: list[ObjectInWorld] or None
+            list containing ObjectInWorld objects, containing position and covariance value. None if conditions not met
+            and model not ran
+        """
+        # Store new input data
+        self.__current_bucket += self.__convert_detections_to_point(detections)
+
+        # Decide to run
+        if not run_override and not self.__decide_to_run():
+            return False, None
+        
+        # Fit points and get cluster data
+        self.reset_model() 
+        self.__vgmm_model = self.__vgmm_model.fit(self.__all_points)
+
+        model_output:list[np.array, float, float] = list(zip(self.__vgmm_model.means_,
+                                                            self.__vgmm_model.weights_,
+                                                            self.__vgmm_model.covariances_))
+        # Empty cluster removal
+        model_output = self.__filter_by_points_ownership(model_output)
+
+        # Sort weights from largest to smallest, along with corresponding covariances and means
+        # TODO: Tuple should come already sorted if model converges, but appears to not come sort in one
+        # of the tests, hence sorting is done every call. Investigate behaviour exactly to only sort when
+        # necessary
+        model_output = self.__sort_by_weights(model_output)
+
+        # Loop through each cluster and drop all clusters after a __WEIGHT_DROP_THRESHOLD drop in weight occured
+        num_viable_clusters = 1
+        weight_drop_thresh_reached = False
+        while num_viable_clusters < len(model_output) and not weight_drop_thresh_reached:
```

Comment:
Seems rather convoluted, why not a for loop:
```py
viable_clusters = [model_output[0]]
for i in range(1, len(model_output)):
    if model_output[i][1] / model_output[i - 1][1] < self.__WEIGHT_DROP_THRESHOLD:
        break

    viable_clusters.append(model_output[i])

Reply:
I only saw this now sorry + implemented. The github website comment system is weird to me and sometimes I can't find all the comments.